File: src/VBNetFx/Controls/ADNullableValueSubControls/ADNullableValueControlBaseClasses.vb
Imports System.Windows.Forms
Imports System.Drawing

Public Interface IADCaptionForNullableValueControl

    Property Alignment() As ContentAlignment
    Property BorderStyle() As BorderStyle
    Property Font() As Font
    Property BackColor() As Color
    Property ForeColor() As Color
    Property Text() As String
    'Liefert ein Control-Array zurück, dass dem umgebenden
    'Steuerelement hinzugefügt wird. Damit kann das Beschriftungs-Control
    'durchaus auch aus mehreren Controls bestehen
    ReadOnly Property Controls() As Control()
    Property Location() As Point
    Property Size() As Size
    'Definiert die Größenreglementierung der Beschriftung
    'Positiver Wert: Mindesthöhe
    'Negativer Wert: Fixe Höhe
    Function MeasureHeight() As Integer
    Property Parent() As Control
    Sub Invalidate()

End Interface

Public Interface IADEditableValueForNullableValueControl

    Property Font() As Font
    Property BackColor() As Color
    Property ForeColor() As Color
    'Der aktuelle Wert als Object, damit
    'ihn jedes belibige Control darstellen kann
    Property Value() As Object
    'Liefert ein Control-Array zurück, dass dem umgebenden
    'Steuerelement hinzugefügt wird. Damit kann das Werte-Control
    'durchaus auch aus mehreren Controls bestehen
    ReadOnly Property Controls() As Control()
    Property Location() As Point
    Property Size() As Size
    Property Parent() As Control

    'Definiert die Größenreglementierung der Beschriftung
    'Positiver Wert: Mindesthöhe
    'Negativer Wert: Fixe Höhe
    Function MeasureHeight() As Integer


    'Wird durch Chance getriggert, während Modified und Modified Changed
    'durch den Entwickler ausschließlich über Modified getriggert wird!
    Property OnceModified() As Boolean
    Sub ResetOnceModified()

    Event ValueChanged(ByVal sender As Object, ByVal e As System.EventArgs)
    Event Validated(ByVal sender As Object, ByVal e As System.EventArgs)
    Event Validating(ByVal sender As Object, ByVal e As System.ComponentModel.CancelEventArgs)
    Event OnceModifiedChanged(ByVal sender As Object, ByVal e As System.EventArgs)

    Sub OnValueChanged(ByVal e As System.EventArgs)
    Sub OnValidated(ByVal e As System.EventArgs)
    Sub OnValidating(ByVal e As System.ComponentModel.CancelEventArgs)
    Sub OnOnceModifiedChanged(ByVal e As System.EventArgs)
    Sub Invalidate()

End Interface

Public MustInherit Class ADCaptionForNullableValueControlTemplate
    Implements IADCaptionForNullableValueControl

    Protected WithEvents myControl As Control
    Protected Shared myInstanceCounter As Integer

    Sub New()
        'Wird in abgeleiteten Controls zur Namensbestimmung herangezoegen
        myInstanceCounter += 1
    End Sub

    Public Overridable Property BackColor() As System.Drawing.Color Implements IADCaptionForNullableValueControl.BackColor
        Get
            Return myControl.BackColor
        End Get
        Set(ByVal Value As System.Drawing.Color)
            myControl.BackColor = Value
        End Set
    End Property

    Public Overridable Property Font() As System.Drawing.Font Implements IADCaptionForNullableValueControl.Font
        Get
            Return myControl.Font
        End Get
        Set(ByVal Value As System.Drawing.Font)
            myControl.Font = Value
        End Set
    End Property

    Public Overridable Property ForeColor() As System.Drawing.Color Implements IADCaptionForNullableValueControl.ForeColor
        Get
            Return myControl.ForeColor
        End Get
        Set(ByVal Value As System.Drawing.Color)
            myControl.ForeColor = Value
        End Set
    End Property

    Public Overridable Property Location() As System.Drawing.Point Implements IADCaptionForNullableValueControl.Location
        Get
            Return myControl.Location
        End Get
        Set(ByVal Value As System.Drawing.Point)
            myControl.Location = Value
        End Set
    End Property

    Public Overridable Property Size() As System.Drawing.Size Implements IADCaptionForNullableValueControl.Size
        Get
            Return myControl.Size
        End Get
        Set(ByVal Value As System.Drawing.Size)
            myControl.Size = Value
        End Set
    End Property

    Public Overridable Property Text() As String Implements IADCaptionForNullableValueControl.Text
        Get
            Return myControl.Text
        End Get
        Set(ByVal Value As String)
            myControl.Text = Value
        End Set
    End Property

    Public Overridable ReadOnly Property Controls() As System.Windows.Forms.Control() Implements IADCaptionForNullableValueControl.Controls
        Get
            Dim locControl(0) As Control
            locControl(0) = myControl
            Return locControl
        End Get
    End Property

    Public Overridable Function MeasureHeight() As Integer Implements IADCaptionForNullableValueControl.MeasureHeight
        Return 0
    End Function

    Public MustOverride Property Alignment() As System.Drawing.ContentAlignment Implements IADCaptionForNullableValueControl.Alignment
    Public MustOverride Property BorderStyle() As System.Windows.Forms.BorderStyle Implements IADCaptionForNullableValueControl.BorderStyle

    Public Property Parent() As System.Windows.Forms.Control Implements IADCaptionForNullableValueControl.Parent
        Get
            Return myControl.Parent
        End Get
        Set(ByVal value As System.Windows.Forms.Control)
            myControl.Parent = value
        End Set
    End Property

    Public Sub Invalidate() Implements IADCaptionForNullableValueControl.Invalidate
        myControl.Invalidate()
    End Sub
End Class

Public MustInherit Class ADEditableValueForNullableValueControlTemplate(Of ValType As IComparable)
    Implements IADEditableValueForNullableValueControl

    Protected myControl As Control
    Protected myValue As ADDBNullable(Of ValType)
    Protected myAssignedValue As ADDBNullable(Of ValType)
    Protected Shared myInstanceCounter As Integer
    Protected myOnceModified As Boolean

    Public Event ValueChanged(ByVal sender As Object, ByVal e As System.EventArgs) Implements IADEditableValueForNullableValueControl.ValueChanged
    Public Event Validated(ByVal sender As Object, ByVal e As System.EventArgs) Implements IADEditableValueForNullableValueControl.Validated
    Public Event Validating(ByVal sender As Object, ByVal e As System.ComponentModel.CancelEventArgs) Implements IADEditableValueForNullableValueControl.Validating
    Public Event OnceModifiedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Implements IADEditableValueForNullableValueControl.OnceModifiedChanged

    Sub New()
        'Alte Überlegung
        myInstanceCounter += 1
    End Sub

    Public Overridable Property BackColor() As System.Drawing.Color Implements IADEditableValueForNullableValueControl.BackColor
        Get
            Return myControl.BackColor
        End Get
        Set(ByVal Value As System.Drawing.Color)
            myControl.BackColor = Value
        End Set
    End Property

    Public Overridable Property Font() As System.Drawing.Font Implements IADEditableValueForNullableValueControl.Font
        Get
            Return myControl.Font
        End Get
        Set(ByVal Value As System.Drawing.Font)
            myControl.Font = Value
        End Set
    End Property

    Public Overridable Property ForeColor() As System.Drawing.Color Implements IADEditableValueForNullableValueControl.ForeColor
        Get
            Return myControl.ForeColor
        End Get
        Set(ByVal Value As System.Drawing.Color)
            myControl.ForeColor = Value
        End Set
    End Property

    Public Overridable Property Location() As System.Drawing.Point Implements IADEditableValueForNullableValueControl.Location
        Get
            Return myControl.Location
        End Get
        Set(ByVal Value As System.Drawing.Point)
            myControl.Location = Value
        End Set
    End Property

    Public Overridable Property Size() As System.Drawing.Size Implements IADEditableValueForNullableValueControl.Size
        Get
            Return myControl.Size
        End Get
        Set(ByVal Value As System.Drawing.Size)
            myControl.Size = Value
        End Set
    End Property

    Public Overridable Property OnceModified() As Boolean Implements IADEditableValueForNullableValueControl.OnceModified
        Get
            Return myOnceModified
        End Get
        Set(ByVal Value As Boolean)
            If Value <> myOnceModified And Value = True Then
                OnOnceModifiedChanged(EventArgs.Empty)
            End If
            myOnceModified = Value
        End Set
    End Property

    Public MustOverride Property Value() As Object Implements IADEditableValueForNullableValueControl.Value

    Public Overridable ReadOnly Property Controls() As System.Windows.Forms.Control() Implements IADEditableValueForNullableValueControl.Controls
        Get
            Dim locControl(0) As Control
            locControl(0) = myControl
            Return locControl
        End Get
    End Property

    'Bestimmt die Höhe eines Controls
    'negativer Wert: Fixe Höhe
    'positivert Wert: Mindesthöhe
    '0: keine Festlegung
    Public Overridable Function MeasureHeight() As Integer Implements IADEditableValueForNullableValueControl.MeasureHeight
        Return 0
    End Function

    Public Overridable Sub OnValueChanged(ByVal e As System.EventArgs) Implements IADEditableValueForNullableValueControl.OnValueChanged
        RaiseEvent ValueChanged(Me, e)
        OnceModified = OnceModified Or True
    End Sub

    Public Overridable Sub OnValidated(ByVal e As System.EventArgs) Implements IADEditableValueForNullableValueControl.OnValidated
        RaiseEvent Validated(Me, e)
    End Sub

    Public Overridable Sub OnValidating(ByVal e As System.ComponentModel.CancelEventArgs) Implements IADEditableValueForNullableValueControl.OnValidating
        RaiseEvent Validating(Me, e)
    End Sub

    Public Overridable Sub OnOnceModifiedChanged(ByVal e As System.EventArgs) Implements IADEditableValueForNullableValueControl.OnOnceModifiedChanged
        RaiseEvent OnceModifiedChanged(Me, e)
    End Sub

    Public Overridable Sub ResetOnceModified() Implements IADEditableValueForNullableValueControl.ResetOnceModified
        'Löst kein Ereignis aus! Sie können aber OnceModified überschreiben,
        'falls Sie dieses Ereignis dennoch in einen Event umwandeln möchten
        OnceModified = False
    End Sub

    Public Property Parent() As System.Windows.Forms.Control Implements IADEditableValueForNullableValueControl.Parent
        Get
            Return myControl.Parent
        End Get
        Set(ByVal value As System.Windows.Forms.Control)
            myControl.Parent = value
        End Set
    End Property

    Public Sub Invalidate() Implements IADEditableValueForNullableValueControl.Invalidate
        myControl.Invalidate()
    End Sub
End Class

File: src/NetFxCSharp/Controls/ADNullableValueSubControls/ADNullableValueControlBaseClasses.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace ActiveDev.Controls
{
    public interface IADCaptionForNullableValueControl
    {
        ContentAlignment Alignment { get; set; }

        BorderStyle BorderStyle { get; set; }

        Font Font { get; set; }

        Color BackColor { get; set; }

        Color ForeColor { get; set; }

        string Text { get; set; }

        //Liefert ein Control-Array zurück, dass dem umgebenden
        //Steuerelement hinzugefügt wird. Damit kann das Beschriftungs-Control
        //durchaus auch aus mehreren Controls bestehen
        Control[] Controls { get; }

        Point Location { get; set; }

        Size Size { get; set; }

        //Definiert die Größenreglementierung der Beschriftung
        //Positiver Wert: Mindesthöhe
        //Negativer Wert: Fixe Höhe
        int MeasureHeight();
        Control Parent { get; set; }

        void Invalidate();
    }

    public interface IADEditableValueForNullableValueControl
    {
        Font Font { get; set; }

        Color BackColor { get; set; }

        Color ForeColor { get; set; }

        //Der aktuelle Wert als Object, damit
        //ihn jedes belibige Control darstellen kann
        object Value { get; set; }

        //Liefert ein Control-Array zurück, dass dem umgebenden
        //Steuerelement hinzugefügt wird. Damit kann das Werte-Control
        //durchaus auch aus mehreren Controls bestehen
        Control[] Controls { get; }

        Point Location { get; set; }

        Size Size { get; set; }

        Control Parent { get; set; }

        //Definiert die Größenreglementierung der Beschriftung
        //Positiver Wert: Mindesthöhe
        //Negativer Wert: Fixe Höhe
        int MeasureHeight();
        //Wird durch Chance getriggert, während Modified und Modified Changed
        //durch den Entwickler ausschließlich über Modified getriggert wird!
        bool OnceModified { get; set; }

        void ResetOnceModified();
        delegate void ValueChangedEventHandler(object sender, System.EventArgs e);
        event ValueChangedEventHandler ValueChanged;
        delegate void ValidatedEventHandler(object sender, System.EventArgs e);
        event ValidatedEventHandler Validated;
        delegate void ValidatingEventHandler(object sender, System.ComponentModel.CancelEventArgs e);
        event ValidatingEventHandler Validating;
        delegate void OnceModifiedChangedEventHandler(object sender, System.EventArgs e);
        event OnceModifiedChangedEventHandler OnceModifiedChanged;
        void OnValueChanged(System.EventArgs e);
        void OnValidated(System.EventArgs e);
        void OnValidating(System.ComponentModel.CancelEventArgs e);
        void OnOnceModifiedChanged(System.EventArgs e);
        void Invalidate();
    }

    public abstract class ADCaptionForNullableValueControlTemplate : IADCaptionForNullableValueControl
    {
        protected Control myControl;
        protected static int myInstanceCounter;
        public ADCaptionForNullableValueControlTemplate()
        {
            //Wird in abgeleiteten Controls zur Namensbestimmung herangezoegen
            myInstanceCounter += 1;
        }

        public virtual System.Drawing.Color BackColor
        {
            get
            {
                return myControl.BackColor;
            }

            set
            {
                System.Drawing.Color Value = value;
                myControl.BackColor = Value;
            }
        }

        public virtual System.Drawing.Font Font
        {
            get
            {
                return myControl.Font;
            }

            set
            {
                System.Drawing.Font Value = value;
                myControl.Font = Value;
            }
        }

        public virtual System.Drawing.Color ForeColor
        {
            get
            {
                return myControl.ForeColor;
            }

            set
            {
                System.Drawing.Color Value = value;
                myControl.ForeColor = Value;
            }
        }

        public virtual System.Drawing.Point Location
        {
            get
            {
                return myControl.Location;
            }

            set
            {
                System.Drawing.Point Value = value;
                myControl.Location = Value;
            }
        }

        public virtual System.Drawing.Size Size
        {
            get
            {
                return myControl.Size;
            }

            set
            {
                System.Drawing.Size Value = value;
                myControl.Size = Value;
            }
        }

        public virtual string Text
        {
            get
            {
                return myControl.Text;
            }

            set
            {
                string Value = value;
                myControl.Text = Value;
            }
        }

        public virtual System.Windows.Forms.Control[] Controls
        {
            get
            {
                Control[] locControl = new Control[1];
                locControl[0] = myControl;
                return locControl;
            }
        }

        public virtual int MeasureHeight()
        {
            return 0;
        }

        public abstract System.Drawing.ContentAlignment Alignment { get; set; }
        public abstract System.Windows.Forms.BorderStyle BorderStyle { get; set; }

        public System.Windows.Forms.Control Parent
        {
            get
            {
                return myControl.Parent;
            }

            set
            {
                myControl.Parent = value;
            }
        }

        public void Invalidate()
        {
            myControl.Invalidate();
        }
    }

    public abstract class ADEditableValueForNullableValueControlTemplate<ValType> : IADEditableValueForNullableValueControl where ValType : IComparable
    {
        protected Control myControl;
        protected ADDBNullable<ValType> myValue;
        protected ADDBNullable<ValType> myAssignedValue;
        protected static int myInstanceCounter;
        protected bool myOnceModified;
        public event ActiveDev.Controls.IADEditableValueForNullableValueControl.ValueChangedEventHandler ValueChanged;
        public event ActiveDev.Controls.IADEditableValueForNullableValueControl.ValidatedEventHandler Validated;
        public event ActiveDev.Controls.IADEditableValueForNullableValueControl.ValidatingEventHandler Validating;
        public event ActiveDev.Controls.IADEditableValueForNullableValueControl.OnceModifiedChangedEventHandler OnceModifiedChanged;
        public ADEditableValueForNullableValueControlTemplate()
        {
            //Alte Überlegung
            myInstanceCounter += 1;
        }

        public virtual System.Drawing.Color BackColor
        {
            get
            {
                return myControl.BackColor;
            }

            set
            {
                System.Drawing.Color Value = value;
                myControl.BackColor = Value;
            }
        }

        public virtual System.Drawing.Font Font
        {
            get
            {
                return myControl.Font;
            }

            set
            {
                System.Drawing.Font Value = value;
                myControl.Font = Value;
            }
        }

        public virtual System.Drawing.Color ForeColor
        {
            get
            {
                return myControl.ForeColor;
            }

            set
            {
                System.Drawing.Color Value = value;
                myControl.ForeColor = Value;
            }
        }

        public virtual System.Drawing.Point Location
        {
            get
            {
                return myControl.Location;
            }

            set
            {
                System.Drawing.Point Value = value;
                myControl.Location = Value;
            }
        }

        public virtual System.Drawing.Size Size
        {
            get
            {
                return myControl.Size;
            }

            set
            {
                System.Drawing.Size Value = value;
                myControl.Size = Value;
            }
        }

        public virtual bool OnceModified
        {
            get
            {
                return myOnceModified;
            }

            set
            {
                bool Value = value;
                if (Value != myOnceModified & Value == true)
                {
                    OnOnceModifiedChanged(EventArgs.Empty);
                }

                myOnceModified = Value;
            }
        }

        public abstract object Value { get; set; }

        public virtual System.Windows.Forms.Control[] Controls
        {
            get
            {
                Control[] locControl = new Control[1];
                locControl[0] = myControl;
                return locControl;
            }
        }

        //Bestimmt die Höhe eines Controls
        //negativer Wert: Fixe Höhe
        //positivert Wert: Mindesthöhe
        //0: keine Festlegung
        public virtual int MeasureHeight()
        {
            return 0;
        }

        public virtual void OnValueChanged(System.EventArgs e)
        {
            ValueChanged?.Invoke(this, e);
            OnceModified = OnceModified | true;
        }

        public virtual void OnValidated(System.EventArgs e)
        {
            Validated?.Invoke(this, e);
        }

        public virtual void OnValidating(System.ComponentModel.CancelEventArgs e)
        {
            Validating?.Invoke(this, e);
        }

        public virtual void OnOnceModifiedChanged(System.EventArgs e)
        {
            OnceModifiedChanged?.Invoke(this, e);
        }

        public virtual void ResetOnceModified()
        {
            //Löst kein Ereignis aus! Sie können aber OnceModified überschreiben,
            //falls Sie dieses Ereignis dennoch in einen Event umwandeln möchten
            OnceModified = false;
        }

        public System.Windows.Forms.Control Parent
        {
            get
            {
                return myControl.Parent;
            }

            set
            {
                myControl.Parent = value;
            }
        }

        public void Invalidate()
        {
            myControl.Invalidate();
        }
    }
}